internal/state/listener.go

```diff
@@ -60,68 +61,115 @@ func newListenerConfiguratorFactory(
 	}
 }
 
-type httpsListenerConfigurator struct {
+type httpListenerConfigurator struct {
 	gateway         *v1beta1.Gateway
 	secretMemoryMgr SecretDiskMemoryManager
 	usedHostnames   map[string]*listener
+	validate        func(gl v1beta1.Listener) []conditions.Condition
+}
+
+func newHTTPListenerConfigurator(gw *v1beta1.Gateway) *httpListenerConfigurator {
```

Comment:
Having two separate constructors that return the same type is confusing. I'm also struggling with the naming here. We are using an `httpListenerConfigurator` for both http and https. I'm worried this could lead to errors and mixups in the future.

Reply:
> Having two separate constructors that return the same type is confusing.

why?

> I'm also struggling with the naming here. We are using an httpListenerConfigurator for both http and https. I'm worried this could lead to errors and mixups in the future.

How about `newHTTPListenerConfiguratorForHTTP`, `newHTTPListenerConfiguratorForHTTPS` ?

Reply:
It's confusing because we have the same type that can be constructed in two different ways. We have a field in an object -- SecretMemoryMgr -- that may or may not be nil. In the future, the https configurator could be configured without the secret memory manager, which would lead to a panic. By setting the `http` and `https` fields of the factory object to the same type we are opening ourselves up to developer error. An `httpListenerConfigurator` could be incorrectly configured for the protocol HTTP, and the compiler would not catch it.

Reply:
> It's confusing because we have the same type that can be constructed in two different ways. We have a field in an object -- SecretMemoryMgr -- that may or may not be nil. In the future, the https configurator could be configured without the secret memory manager, which would lead to a panic. By setting the http and https fields of the factory object to the same type we are opening ourselves up to developer error. An httpListenerConfigurator could be incorrectly configured for the protocol HTTP, and the compiler would not catch it.

Those two constructor functions are there to help the developer avoid creating configurator with invalid state.

> it could be configured without the secret memory manager, which would lead to a panic. By setting the http and https fields of the factory object to the same type we are opening ourselves up to developer error. 

We do have unit tests to catch that in case the developer still makes a mistake. 

Since the configurator is package-private helper type, I don't see much value introducing more safeguards there.

However, let me know if you have some ideas in mind